### ISSUE TITLE
Re-evaluate dependency usage

### ISSUE THREAD

**Issue by kevinburkeshyp** [request/request#2136]
Yesterday a developer unpublished the left-pad module and broke a lot of builds. The lesson for me at least was to be a lot more careful about which dependencies we take. Especially in cases where the code is trivially inline-able, it may make sense to bring the code into the project, rather than rely on a dependency (and have all of `request`'s users rely on that dependency).

Here are some that could be inlined:

- `is-typed-array`: https://github.com/hughsk/is-typedarray/blob/master/index.js
- `caseless`: https://github.com/request/caseless/blob/master/index.js
- `isstream`: https://github.com/rvagg/isstream/blob/master/isstream.js
- `json-stringify-safe`: https://github.com/isaacs/json-stringify-safe/blob/master/stringify.js
- `stringstream` (especially if you ignore the AlignedStringDecoder code) https://github.com/mhart/StringStream/blob/master/stringstream.js

Furthermore `readable-stream` (buried under `bl`) takes dependencies on `isarray` and `util-deprecate`, which seem pretty trivial to copy inline.

**Comment by kevinburkeshyp** [comment]
Sorry - when I say "we", I'm referring to my team

**Comment by simov** [comment]
Thanks for stopping by, however I think that's just pure paranoia :laughing: 

I mean you can't possibly suspect each and every open source maintainer of maliciously trying to ruin your life and projects.

Also I really would like to emphasize on that the author of left-pad got one of his modules stolen by the NPM team without his permissions (though the actual email conversation was not published).

As for whether request is huge and have lots of dependencies - yes! It is. The problem is that lots of people got confused when you publish something too modular that requires some learning curve and cognitive load beyond just using a fancy API :)

As for whether you should have a huge lib folder in your module with everything you can possible imagine in it, probably not.

**Comment by kevinburke** [comment]
> As for whether you should have a huge lib folder in your module with everything you can possible imagine in it, probably not.

Isn't this what `node_modules` is though? It's just acknowledging more of the things that you depend on, explicitly.

**Comment by simov** [comment]
Yep having lots of small dependencies is trendy these days. If the surface of your module is small enough you shouldn't be worrying about having too many dependencies or huge lib folder in it. But then again you build your end user applications and more user friendly wrappers on top of these modules so you end up with lots of dependencies anyway.

**Comment by DavidHenri008** [comment]
Is there some consideration for updating isstream to its latest version?
